File: 11. Container With Most Water/main.cpp
#include <vector>

using std::max;
using std::min;
using std::vector;

class Solution {
public:
    int maxArea(vector<int> &height) {
        int left = 0;
        int right = height.size() - 1;
        int mArea = 0;

        while (left < right) {
            int area = (right - left) * min(height[left], height[right]);
            mArea = max(mArea, area);

            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }

        return mArea;
    }
};

File: 11. Container With Most Water/main.rs
struct Solution;

impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let mut left = 0;
        let mut right = height.len() - 1;
        let mut m_area = 0;

        while left < right {
            let area = (right - left) as i32 * height[left].min(height[right]);
            m_area = m_area.max(area);

            if height[left] < height[right] {
                left += 1;
            } else {
                right -= 1;
            }
        }

        m_area
    }
}
